File: packages/types/src/errors/execution-errors.ts
/**
 * Defining Execution-Related Error Types
 * Define the type of error during execution
 */

import { ExecutionError, ErrorSeverity } from "./base.js";

/**
 * Business Logic Error Types
 *
 * Specialized for business logic related execution errors (e.g. route mismatch, condition not satisfied, etc.)
 * Inherited from ExecutionError to maintain backward compatibility
 *
 * Usage Scenario:
 * - Routing condition evaluation failure
 * - Condition node judgment failure
 * - Loop termination condition exception
 * - Business rule validation failure
 */
export class BusinessLogicError extends ExecutionError {
  constructor(
    message: string,
    public readonly businessContext?: string,
    public readonly ruleName?: string,
    nodeId?: string,
    workflowId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(message, nodeId, workflowId, { ...context, businessContext, ruleName }, cause, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * System implementation error type
 *
 * Used for system-level execution errors (e.g. state management failure, context loss, etc.)
 * Inherited from ExecutionError to maintain backward compatibility
 *
 * Note: For more specific error scenarios, use the following specialized error types:
 * - DependencyInjectionError: Dependency injection failure
 * - StateManagementError: State management failure
 * - AgentCheckpointError: Agent checkpoint operation failed
 * - WorkflowCheckpointError: Workflow checkpoint operation failed
 * - EventSystemError: Event system error
 */
export class SystemExecutionError extends ExecutionError {
  constructor(
    message: string,
    public readonly systemComponent?: string,
    public readonly failurePoint?: string,
    nodeId?: string,
    workflowId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(
      message,
      nodeId,
      workflowId,
      { ...context, systemComponent, failurePoint },
      cause,
      severity,
    );
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Dependency Injection Error Types
 *
 * Specialized for dependency injection related errors
 * Used when required services, components or dependencies are not provided or cannot be resolved
 *
 * Usage Scenario:
 * - Required services are not injected
 * - Dependency resolution failure
 * - Service initialization failure
 */
export class DependencyInjectionError extends ExecutionError {
  constructor(
    message: string,
    public readonly dependencyName: string,
    public readonly requiredBy?: string,
    nodeId?: string,
    workflowId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(message, nodeId, workflowId, { ...context, dependencyName, requiredBy }, cause, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Status Management Error Types
 *
 * Specialized for state management related errors
 * Used when a status read, write, update, or delete operation fails
 *
 * Usage Scenario:
 * - Status read failure
 * - Status write failure
 * - Status synchronization failure
 * - Status consistency error
 */
export class StateManagementError extends ExecutionError {
  constructor(
    message: string,
    public readonly stateType: string,
    public readonly operation: "read" | "write" | "delete" | "update" | "sync",
    public readonly stateKey?: string,
    nodeId?: string,
    workflowId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(
      message,
      nodeId,
      workflowId,
      { ...context, stateType, operation, stateKey },
      cause,
      severity,
    );
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Agent Checkpoint Error Type
 *
 * Specialized for agent loop checkpoint-related errors
 * Used when an agent checkpoint creation, recovery, deletion, or validation operation fails
 *
 * Usage Scenario:
 * - Agent checkpoint creation failure
 * - Agent checkpoint recovery failure
 * - Agent checkpoint deletion failure
 * - Agent checkpoint validation failure
 */
export class AgentCheckpointError extends ExecutionError {
  constructor(
    message: string,
    public readonly operation: "create" | "restore" | "delete" | "validate",
    public readonly checkpointId?: string,
    public readonly agentLoopId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(message, undefined, undefined, { ...context, operation, checkpointId, agentLoopId }, cause, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Workflow Checkpoint Error Type
 *
 * Specialized for workflow/graph checkpoint-related errors
 * Used when a workflow checkpoint creation, recovery, deletion, or validation operation fails
 *
 * Usage Scenario:
 * - Workflow checkpoint creation failure
 * - Workflow checkpoint recovery failure
 * - Workflow checkpoint deletion failure
 * - Workflow checkpoint validation failure
 */
export class WorkflowCheckpointError extends ExecutionError {
  constructor(
    message: string,
    public readonly operation: "create" | "restore" | "delete" | "validate",
    public readonly checkpointId?: string,
    public override readonly nodeId?: string,
    public override readonly workflowId?: string,
    public readonly executionId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(
      message,
      nodeId,
      workflowId,
      { ...context, operation, checkpointId, executionId },
      cause,
      severity,
    );
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Event System Error Types
 *
 * Specialized for event system related errors
 * Used when an event firing, listening, or processing operation fails
 *
 * Usage Scenario:
 * - Failure to fire an event
 * - Event listener registration fails
 * - Event handler execution failure
 * - Event bus error
 */
export class EventSystemError extends ExecutionError {
  constructor(
    message: string,
    public readonly operation: "emit" | "subscribe" | "unsubscribe" | "handle",
    public readonly eventType?: string,
    nodeId?: string,
    workflowId?: string,
    context?: Record<string, unknown>,
    cause?: Error,
    severity?: ErrorSeverity,
  ) {
    super(message, nodeId, workflowId, { ...context, eventType, operation }, cause, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

File: crates/foundation/wf-types/src/errors/execution_errors.rs
use crate::errors::WfError;

pub fn business_logic_error(
    message: impl Into<String>,
    business_context: Option<String>,
    rule_name: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    if let Some(v) = business_context {
        ctx.insert("business_context".into(), serde_json::Value::String(v));
    }
    if let Some(v) = rule_name {
        ctx.insert("rule_name".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}

pub fn system_execution_error(
    message: impl Into<String>,
    system_component: Option<String>,
    failure_point: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    if let Some(v) = system_component {
        ctx.insert("system_component".into(), serde_json::Value::String(v));
    }
    if let Some(v) = failure_point {
        ctx.insert("failure_point".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}

pub fn dependency_injection_error(
    message: impl Into<String>,
    dependency_name: String,
    required_by: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    ctx.insert(
        "dependency_name".into(),
        serde_json::Value::String(dependency_name),
    );
    if let Some(v) = required_by {
        ctx.insert("required_by".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}

pub fn state_management_error(
    message: impl Into<String>,
    state_type: String,
    operation: String,
    state_key: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    ctx.insert("state_type".into(), serde_json::Value::String(state_type));
    ctx.insert("operation".into(), serde_json::Value::String(operation));
    if let Some(v) = state_key {
        ctx.insert("state_key".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}

pub fn agent_checkpoint_error(
    message: impl Into<String>,
    operation: String,
    checkpoint_id: Option<String>,
    agent_loop_id: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    ctx.insert("operation".into(), serde_json::Value::String(operation));
    if let Some(v) = checkpoint_id {
        ctx.insert("checkpoint_id".into(), serde_json::Value::String(v));
    }
    if let Some(v) = agent_loop_id {
        ctx.insert("agent_loop_id".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}

pub fn workflow_checkpoint_error(
    message: impl Into<String>,
    operation: String,
    checkpoint_id: Option<String>,
    execution_id: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    ctx.insert("operation".into(), serde_json::Value::String(operation));
    if let Some(v) = checkpoint_id {
        ctx.insert("checkpoint_id".into(), serde_json::Value::String(v));
    }
    if let Some(v) = execution_id {
        ctx.insert("execution_id".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}

pub fn event_system_error(
    message: impl Into<String>,
    operation: String,
    event_type: Option<String>,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    ctx.insert("operation".into(), serde_json::Value::String(operation));
    if let Some(v) = event_type {
        ctx.insert("event_type".into(), serde_json::Value::String(v));
    }
    WfError::new(message).with_context(ctx)
}
